### DISCUSSION TITLE
Incorrect type on custom Eloquent collection `groupBy()`

### DISCUSSION THREAD

**Discussion by oli-laban** [larastan/larastan#2088] [Q&A] [answered]
I have the following custom eloquent collection:

```php
<?php

namespace App\Collections;

use App\Models\TicketPrice;
use Illuminate\Database\Eloquent\Collection;

/** @extends Collection<array-key, TicketPrice> */
final class TicketPriceCollection extends Collection
{
    public function priceBandsWithMinMaxPrice(): PriceBandCollection
    {
        return new PriceBandCollection(
            $this
                ->groupBy('priceBand.id')
                ->map(function (TicketPriceCollection $ticketPrices) {
                    $prices = $ticketPrices->pluck('price');

                    $priceBand = $ticketPrices->first()->priceBand;

                    $priceBand->setAttribute('min_price', $prices->min());
                    $priceBand->setAttribute('max_price', $prices->max());

                    return $priceBand;
                })
                ->all()
        );
    }
}
```

Which is triggering this PHPStan error:

```
Parameter #1 $callback of method
         Illuminate\Database\Eloquent\Collection<(int|string),App\Models\TicketPrice>::map()
         expects callable(App\Models\TicketPrice, int|string): (App\Models\PriceBand|null),
         Closure(App\Collections\TicketPriceCollection): (App\Models\PriceBand|null) given.
```

The return type of the `groupBy()` call is a normal `TicketPriceCollection`:

```php
$groupedPriceBands = $this->groupBy('priceBand.id');
\PHPStan\dumpType($groupedPriceBands);

// App\Collections\TicketPriceCollection
```

When the actual value is a `TicketPriceCollection` _of_ `TicketPriceCollection`s:

```php
$groupedPriceBands = $this->groupBy('priceBand.id');
dd($groupedPriceBands);

// App\Collections\TicketPriceCollection {#1373 ▼ // app/Collections/TicketPriceCollection.php:14
//   #items: array:1 [▼
//     264 => App\Collections\TicketPriceCollection {#1391 ▼
//       #items: array:7 [...]
//     }
//   ]
// }
```

I'm sure it's probably something to do with the PHPDoc for the collection or something. Any help is appreciated.

**Comment by calebdw** [comment]
What do your generics look like for the `TicketPriceCollection` and `PriceBandCollection` collections

**Reply by oli-laban** [reply]
You can see the generic for the `TicketPriceCollection` above. The `PriceBandCollection` is basically the same thing:

```php
/** @extends Collection<array-key, PriceBand> */
final class PriceBandCollection extends Collection
{
    // ...
}
```

**Reply by rudiedirkx** [reply]
Yeah, non-generic collections in my app work funky too. You can even have a slightly-less-generic collection that's still somewhat generic: `@extends Collection<TKey, PriceBand>`. Should be possible, right? Or the opposite: `@extends Collection<array-key, TModel>`.

**Reply by calebdw** [reply, marked as answer]
There lies the problem---both of these collection can only contain a certain model---you cannot call `groupBy` on these collections because they return `static<int, static<...>>` (a collection of collections which violate the generics that you have defined).

You must convert to the base SupportCollection before calling `groupBy`:

```php
            $this
                ->toBase()
                ->groupBy('priceBand.id')
```

**Reply by oli-laban** [reply]
@calebdw Thanks, had no idea `toBase()` existed. I just had to remove the type from the `map()` callback and manually type it with `/* @var Collection<int, TicketPrice> $ticketPrices */` as well (not sure if there's a better way to do that).

**Reply by rudiedirkx** [reply]
If you `toBase()` before `groupBy()` you lose the original collection in the `map()`. Laravel's `groupBy()` should group better (with 2 different collections, outside and inside) but it doesn't, so you can't `toBase()` if you want a `TicketPriceCollection` inside `map()` (which I always do). Since Laravel breaks the generics-contract, shouldn't Larastan just follow that? Or is that impossible because it shouldn't be possible, even though it is?

**Reply by rudiedirkx** [reply]
Larastan even puts in the effort to understand `groupBy()` output in `EnumerableGenericStaticMethodDynamicMethodReturnTypeExtension`. Except exact `EloquentCollection` is excluded, but not its children.... That's probably a Larastan bug.
